src/Microsoft.Kiota.Cli.Commons/Http/Headers/BaseHeadersStore.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.Kiota.Cli.Commons.Http.Headers;
+
+/// <summary>
+/// Base class for the <see cref="IHeadersStore"/> interface.
+/// This type exists so all child classes can share the headers parsing logic.
+/// </summary>
+public abstract class BaseHeadersStore : IHeadersStore
+{
+    /// <inheritdoc />
+    public abstract IEnumerable<KeyValuePair<string, ICollection<string>>> GetHeaders();
+
+    /// <inheritdoc />
+    /// <remarks> 
+    /// Passing a null or empty collection of headers to this function has the
+    /// same effect as calling <see cref="Drain"/>
+    /// </remarks>
+    public virtual IEnumerable<KeyValuePair<string, ICollection<string>>> SetHeaders(ICollection<string>? headers)
+    {
+        var existing = Drain();
+        AddHeaders(headers);
+        return existing;
+    }
+
+    /// <inheritdoc />
+    /// <remarks>
+    /// If <paramref name="headers"/> collection is null, this function will
+    /// return false
+    /// The parsing logic for this type uses the <see cref="ParseHeaders"/>
+    /// function to parse the headers. Override ParseHeaders in sub-classes
+    /// to change the parsing behavior.
+    /// </remarks>
+    public virtual bool AddHeaders(ICollection<string>? headers)
+    {
+        if (headers is null || headers.Count < 1)
+        {
+            return false;
+        }
+
+        var m = ParseHeaders(headers);
+
+        foreach (var (key, value) in m)
+        {
+            DoAddHeader(key, value);
+        }
+
+        return true;
+    }
+
+    /// <inheritdoc />
+    public abstract IEnumerable<KeyValuePair<string, ICollection<string>>> Drain();
+
+    /// <summary>
+    /// Parses a collection of headers into a collection of key-value pairs
+    /// with the header name and value.
+    /// This function expects each header item to be in the format
+    /// <code>header-name=header-value</code>
+    /// This function does not do anything about duplicates, so if you passed
+    /// in <code>["a=1", "b=2"]</code> the result will be
+    /// <code>[{ "a": "1" }, { "b": "2" }]</code>
+    /// </summary>
+    /// <param name="headers">A collection of strings with header information.</param>
+    /// <returns>A collection of key-value pairs with parsed header information.</returns>
+    /// <remarks>Override this function to customize the parsing logic.</remarks>
+    protected virtual IEnumerable<KeyValuePair<string, string>> ParseHeaders(IEnumerable<string> headers)
+    {
+        // This function is called by AddHeaders which checks for null. If
+        // headers is null, then something went wrong
+        Debug.Assert(headers is not null);

```

Comment:
it should probably guard against null instead

Reply:
again, this method is publicly accessible (protected) and could be used in a different context you don't have control over. Yes it's redundant if you look at the whole stack we own, but it provides a better experience than getting random NullReferenceExceptions that point to code location the consumer doesn't own.

Reply:
I disagree. Let's look at what a user has to do to trigger a null reference exception:

1. Turn off nullable compilation option
2. Override the BaseHeaderStore with a non-compliant option
3. Call this function from the override with null
4. Build in release mode only without testing

This is a deliberate misuse of the contract. The .NET Framework has many such cases as well where you shouldn't pass a null to a function and the checks are Asserts.

Reply:
For examples, see https://github.com/dotnet/runtime/blob/3ef2c804b4fea8c94c9bd5cc39f79525361a03d2/src/libraries/Common/src/System/Net/ArrayBuffer.cs#L34
https://github.com/dotnet/runtime/blob/3ef2c804b4fea8c94c9bd5cc39f79525361a03d2/src/libraries/System.Linq/src/System/Linq/Take.cs#L46

Reply:
> This is a deliberate misuse of the contract.

Is attributing intent to consumers when they could end up in that (narrow) edge case simply by mistake. Our responsibility is to guard against those mistakes when the solution is trivial to provide a better experience. Introducing behavior that changes between debug and release is building an "easy trap" for consumers to get surprised at runtime.

I think the disagreement on what the best approach would be in this case also comes from the fact this method is protected as:
- public would clearly mandate defensive programing (the edge case is less narrow)
- internal/private would give us entire use of the method, allowing us to make assumptions.

The whole reason this method is protected is because of the multiple levels of abstractions I'm not sure we or consumers will ever need.
Can you expand on why the split between in memory and base? (besides reusability of parts of the code) and the scenarios where you see people "storing" the headers values anywhere else than in memory?
Would you be open to merging those two classes?

Reply:
I think the debug assert is defensive programming. This is not a general-purpose storage object, it has a specific use case.

Reply:
I'll add a documentation comment.

Reply:
For the extensibility scenario they can always reimplement the interface.

Right now the only thing the division between in memory store and base store is allowing is reusability of roughly two loops. It's not a lot of value added for quite some complexity and public API surface we'll have to maintain over time.

Hence my recommendation, merge them together, it'll reduce the complexity, we'll be able to make safe assumptions regarding nullability of things, and people can simply implement the interface should they want different behaviour.

Reply:
I believe that would complicate the extensibility story. If we wanted to make the extension ourselves for new use-cases, we'd also need to duplicate the code too and add more maintenance for ourselves as well. I want it to be customizable because I feel the feature is still kind of MVP. If we GA with a poor extensibility, adding features would be hard.

Reply:
Alright, we've arrived to a stale mate, looping in @andrueastman and @rkodev to provide additional votes and unlock the situation.
You can go ahead and read the whole thread here, I'm also going to try to recap it as impartially as possible:
- Caleb wants to enable maximum reusability and extensibility of the code through: an interface, a base class and a dedicated implementation. He also doesn't believe we should systematically check for nullability in public/protected API surfaces (defensive programming) as this sometimes leads to redundant code.
- I disagree with the lack of defensive programming that behaves consistently in release/debug when we provide public/protected APIs and first suggested to add it even if that introduces a bit of redundancy since NRT do not guarantee null values cannot be passed. Given the pushback on that solution, the presence of an interface, and the limited code being reused in the base class, I suggested merging the base and implementation to circumvent the issue all together.

Reply:
Just an addition on @baywet's 1st point, I agreed with checks on the public API functions and added debug asserts in the protected functions. The invariants are already documented as part of the code docs. The disagreement isn't about defensive programming vs absolutely no defensive programming, it's in what counts as defensive programming. I think your point is Debug.Asserts don't count which I disagree with.

Reply:
With regards to using Debug asserts, I believe we should at least an `ArgumentException.ThrowIfNull` or equivalent statement in the function to check for null. At the end of the day, the current `ParseHeaders` is public/externally visible through an extensibility point out of our control which we can't really make assurances for. If someone builds something on top of it and an unpredicted input/situation occurs in a production/release environment (a case the developer missed in testing where some other dependent code passed a null value), the error will be explicit and easy to identify as opposed to a `NullReferenceException` which can be vague. (Alternatively, the function can simply be made private as it's not part of the `IHeadersStore` contract/interface and let the work of parsing headers be the responsibility of the `AddHeaders` function. Or should it be part of the interface if it's also an extensibility point?)

With regards to extensibility, I'd like to believe that there is a middle ground where we can achieve the same by using **either** abstract class or interface to enforce the extensibility contracts but not necessarily both. The abstract base class may be enough as the extension point can be changed while keeping the defaults provided by the base. This saves time for the extender who only changes what they need rather than re-implementing everything (as currently, we provide pretty good default implementations 😄 ).